pyccel/ast/basic.py

```diff
@@ -165,7 +165,92 @@ def get_attribute_nodes(self, search_type, excluded_nodes = ()):
         self._recursion_in_progress = False
         return results
 
-    def substitute(self, original, replacement, excluded_nodes = ()):
+    def is_attribute_of(self, search_node, excluded_nodes = ()):
+        """ Identifies whether the object search_node is an attribute of
+        this object or one of its attributes
+
+        Parameters
+        ----------
+        search_node : Basic
+                      The object which we are looking for
+        excluded_nodes : tuple of types
+                      Types for which get_user_nodes should not be called
+
+        Results
+        -------
+        bool
+        """
+        if self._recursion_in_progress or len(self._user_nodes) == 0:
+            return []
+        else:
+            self._recursion_in_progress = True
+
+            if any(p is search_node for p in self._user_nodes):
+                self._recursion_in_progress = False
+                return True
+
+            for p in self._user_nodes:
+                if not self.ignore(p) and not isinstance(p, excluded_nodes):
+                    res = p.is_attribute_of(search_node, excluded_nodes = excluded_nodes)
+                    if res:
+                        self._recursion_in_progress = False
+                        return True
+
+            self._recursion_in_progress = False
+            return False
+
+    def is_user_of(self, search_node, excluded_nodes = ()):
+        """ Identifies whether the object search_node is a user
+        of this object. The function searches recursively up the user tree
```

Comment:
Also here we need to change the docstring:
```suggestion
    def is_user_of(self, search_node, excluded_nodes = ()):
        """ Identifies whether this object is a user of search_node.
        The function searches recursively up the user tree
```